modules: mcuboot: enable support for RAMLOAD mode with revert

danieldegrasse opened the pull request:
This PR adds support for RAMLOAD mode with revert in MCUBoot, and enables a testcase for it with the mcumgr subsystem. RAMLOAD mode with revert functions similar to the direct xip mode with revert- MCUBoot will not ramload an image that is not confirmed or marked pending.

~This PR also modifies the method of getting the active slot number when using ramload modes- we need to query the *flash area ID*, not the active boot slot for use with mcuboot boot utilities~

danieldegrasse committed:
samples: subsys: mcumgr: rework ram load support for multiple boards

MCUMgr ram load support was hardcoding the overlay file for the
nrf52840dk. To make this support generic, move the _ram_load overlay
file to the MCUBoot repository, as part of the mcuboot application.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
soc: nxp: common: disable FLASH_MCUX_FLEXSPI_XIP if using ramload

MCUBoot RAMLOAD mode relies on CONFIG_XIP=n, but FLASH_MCUX_FLEXSPI_XIP
y-selects this symbol. Disable CONFIG_FLASH_MCUX_FLEXSPI_XIP for the
case where we are using MCUBoot ramload mode.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
samples: subsys: mgmt: mcumgr: add support for ramload on mimxrt1050_evk

Add support for using the ramload mode of MCUBoot on the mimxrt1050_evk

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
modules: mcuboot: add Kconfigs for RAMLOAD_WITH_REVERT mode

Add Kconfigs for RAMLOAD_WITH_REVERT mode in MCUBoot. This mode works in
a manner similar to DIRECT_XIP_WITH_REVERT- namely, mcuboot will only
boot an image that is either confirmed or marked as pending. If both
images are confirmed, mcuboot will still select the one with the higher
version, so downgrading is not possible using this mode.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
cmake: mcuboot: respect alignment size when signing for ramload mode

When signing for ramload mode, respect the write alignment size setting.
This is required when creating a confirmed image, as the BOOT_MAGIC
value changes based on the alignment size in use.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
cmake: mcuboot: prefer confirmed image when one is created

Prefer flashing the confirmed image when one is created with
CONFIG_MCUBOOT_GENERATE_CONFIRMED_IMAGE. This way, the west runner will
flash this image over the signed image if it exists.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
soc: nxp: flexspi_xip: Update FLASH_MCUX_FLEXSPI_XIP ramload dependencies

CONFIG_FLASH_MCUX_FLEXSPI_XIP should also be disabled when using MCUBoot
ramload mode with revert support.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
dfu: mcuboot: update dependencies for RAMLOAD_WITH_REVERT

Much like in RAMLOAD mode, RAMLOAD_WITH_REVERT requires that mcuboot
subsystem fetch bootloader information via the retention subsystem.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
dfu: boot: mcuboot: fix boot_fetch_active_slot

boot_fetch_active_slot needs to map the slot number to a flash ID, as
this is what the DFU subsystem expects when interacting with the flash
partition.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
samples: mgmt: mcumgr: add testcase for ramload with revert

Add testcase to build mcuboot and smp_svr application using ramload
with revert mode

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
scripts: ci: exempt BOOT_RAM_LOAD_REVERT from checks

Add BOOT_RAM_LOAD_REVERT as a defined Kconfig to the whitelist

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse commented:
Note: this PR requires https://github.com/mcu-tools/mcuboot/pull/2197 from MCUBoot. It also contains commits from https://github.com/zephyrproject-rtos/zephyr/pull/85096, because I validated this support on the RT1050-EVK

danieldegrasse commented:
> This is not working. Do a build with sysbuild and flash using `west flash` so you have both images. If you run the image info command you get the following:
> 
> ```
> Slot 0:
>     Active
>     Confirmed
>     Bootable
> 
> Slot 1:
>     Bootable
>     Pending
>     Permanent
> ```
> 
> This state is invalid as when you reboot, it boots slot 0 again. You also cannot erase the update slot or transfer a new image because it thinks an upgrade is in progress when it is not

How are you testing this? The sequence I'm using is the following:
- build using sysbuild, and flash. MCUBoot and image 0 are flashed.
- rebuild sample with `CONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION="0.0.1+0"`, so the new SMP SVR application has a higher version
- upload the new `zephyr.signed.bin` to slot1 using mcumgr-client
- Mark the image as pending using mcumgr-client, and reset the board
- Verify that the new smp_svr application has booted, and that it will be reverted after a reboot (since it did not confirm itself)

nordicjm commented:
Seems I did not do recover between tests, reproduction is build and flash, then upload the slot1 image **without** changing version, upgrade is bricked as image flag does nothing, new image is not booted, application cannot erase image because it is persistently marked as test

nordicjm commented:
Then after, if you actually do erase the flash of the upgrade slot header, MCUboot shows no secondary image - good, MCUmgr shows no secondary image - good, and try to upload, it fails to upload because of "There is no free slot to place the image" ??? Assuming this is because the flags are still set in the main image but this also bricks updates

nordicjm commented:
Actually a bit confused why a slot1 bin file is generated since the slot0 and slot1 bin files would be exactly the same for RAM load, hex makes sense but don't need the bin for slot1

danieldegrasse commented:
> Seems I did not do recover between tests, reproduction is build and flash, then upload the slot1 image **without** changing version, upgrade is bricked as image flag does nothing, new image is not booted, application cannot erase image because it is persistently marked as test

Trying to reproduce this on my end- the RT1050 build only has bin files, so I'm doing something like the following:
- flash normally using `west flash` (populates slot0 with a confirmed image, and loads MCUBoot)
- load secondary confirmed image using `JLinkExe` at offset for slot1 partition (loading `zephyr.slot1.signed.confirmed.bin`, using `loadfile build/smp_svr/zephyr/zephyr.slot1.signed.confirmed.bin 0x603c0000`
- reset board, verify image0 still boots and image1 is present
- Rebuild with ` -DCONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION=\"0.0.1+0\"` to set new version
- Upload image to slot 1 using `mcumgr-client`: `mcumgr-client -d /dev/ttyACM0 upload build/smp_svr/zephyr/zephyr.signed.bin`
- Mark the image for test `mcumgr-client -d /dev/ttyACM0 test 8fe10b45d6b3eae2183c36e11caf736dbe558db8494ab238458bfcd70e5c092b`
- Reset board, verify image boots once than is removed on next boot (since it doesn't confirm itself)

Are these reproduction steps right? Not sure if I'm missing something here

nordicjm commented:
Build and flash, use a program to upload the secondary image without changing version (not jlink, upload using MCUmgr), mark it as test/confirmed, reboot, device state is stuck

danieldegrasse commented:
> Build and flash, use a program to upload the secondary image without changing version (not jlink, upload using MCUmgr), mark it as test/confirmed, reboot, device state is stuck

Really not sure what I'm missing, but I still don't see an issue. These are the build commands I'm running, after erasing the flash of the RT1050:
```
$ west build -p always -b mimxrt1050_evk/mimxrt1052/hyperflash -T samples/subsys/mgmt/mcumgr/smp_svr/sample.mcumgr.smp_svr.ram_load_revert.serial -- -DCONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION=\"0.0.0+0\"
$ west flash -r jlink
# Rebuild so the image sha will actually be different (otherwise the image cannot be marked for test)
$ west build -p always -b mimxrt1050_evk/mimxrt1052/hyperflash -T samples/subsys/mgmt/mcumgr/smp_svr/sample.mcumgr.smp_svr.ram_load_revert.serial -- -DCONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION=\"0.0.0+0\"
# Upload new image
$ ../../mcumgr-client/target/release/mcumgr-client -d /dev/ttyACM0 upload build/smp_svr/zephyr/zephyr.signed.bin
# Mark image for test
$ ../../mcumgr-client/target/release/mcumgr-client -d /dev/ttyACM0 test 69732fb7b5dbcf0b9c7f3c62ad073894b631093d2045fa9b36204b0b6c50432e
# Reset board, first slot image still boots (since version numbers are equivalent)
# Build new image with updated version
$  west build -p always -b mimxrt1050_evk/mimxrt1052/hyperflash -T samples/subsys/mgmt/mcumgr/smp_svr/sample.mcumgr.smp_svr.ram_load_revert.serial -- -DCONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION=\"0.0.1+0\"
$ ../../mcumgr-client/target/release/mcumgr-client -d /dev/ttyACM0 upload build/smp_svr/zephyr/zephyr.signed.bin
# Mark new image for test
$ ../../mcumgr-client/target/release/mcumgr-client -d /dev/ttyACM0 test 8c21e43c3a1ef4d9b5aca59da6ec79a6bd18c078de8a006d49326082d540828f
# Reset board, verify new image is running
# Reset board again, verify that the image is removed from flash as it did not confirm itself
```

This all seems functional to me. Do these steps not work for you? Or is the concern that the new image with the same version number does not run? I think that is expected behavior- the only way to indicate which image mcuboot should select to run with this feature is via the version number

nordicjm commented:
Build sample with `CONFIG_MCUMGR_GRP_IMG_ALLOW_ERASE_PENDING=n`, flash to device, then **without changing any Kconfigs or anything or rebuilding** just upload the slot1 image using MCUmgr with any supported tool from file `zephyr.slot1.signed.confirmed.bin` and reboot - at this point the device will boot from slot 0 **not slot 1 as it should according to the flags**, then try uploading an image or erasing the slot using MCUmgr, you will get an error that there is no free slot to place the image, and your device is "bricked" because you can no longer load a firmware update to it (without erasing the flash with a debugger, that is)

danieldegrasse committed:
mgmt: update img_mgmt_slot_in_use for ramload with revert mode

We should not block erasing pending images when using ramload with
revert mode, because uploading multiple confirmed images with the same
version would brick the device (preventing future FW updates). Update
the dependencies of img_mgmt_slot_in_use to account for this.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse commented:
> Build sample with `CONFIG_MCUMGR_GRP_IMG_ALLOW_ERASE_PENDING=n`, flash to device, then **without changing any Kconfigs or anything or rebuilding** just upload the slot1 image using MCUmgr with any supported tool from file `zephyr.slot1.signed.confirmed.bin` and reboot - at this point the device will boot from slot 0 **not slot 1 as it should according to the flags**, then try uploading an image or erasing the slot using MCUmgr, you will get an error that there is no free slot to place the image, and your device is "bricked" because you can no longer load a firmware update to it (without erasing the flash with a debugger, that is)

Ahh ok- now this makes sense, thank you- I've made the following change in this push, should fix it. I believe we just want to always allow erasing a pending/confirmed image if it is inactive- otherwise the user can flash two images that are both confirmed, and effectively brick their device

```diff
diff --git a/subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c b/subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c
index 5333d651c9f..1860269d87f 100644
--- a/subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c
+++ b/subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c
@@ -332,7 +332,8 @@ img_mgmt_slot_in_use(int slot)
        int active_slot = img_mgmt_active_slot(image);

 #if !defined(CONFIG_MCUBOOT_BOOTLOADER_MODE_DIRECT_XIP) && \
-       !defined(CONFIG_MCUBOOT_BOOTLOADER_MODE_RAM_LOAD)
+       !defined(CONFIG_MCUBOOT_BOOTLOADER_MODE_RAM_LOAD) && \
+       !defined(CONFIG_MCUBOOT_BOOTLOADER_MODE_RAM_LOAD_WITH_REVERT)
        enum img_mgmt_next_boot_type type = NEXT_BOOT_TYPE_NORMAL;
        int nbs = img_mgmt_get_next_boot_slot(image, &type);
```

danieldegrasse commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -100,8 +107,46 @@ uint8_t boot_fetch_active_slot(void)
+		fa_id = FIXED_PARTITION_ID(SLOT0_PARTITION);
+		break;
+
+	case 1:
+		fa_id = FIXED_PARTITION_ID(SLOT1_PARTITION);
+		break;
```
Added- thanks. What will we need to do on the MCUBoot side to synchronize this? I think this change will have to go in when we update MCUBoot to include https://github.com/mcu-tools/mcuboot/pull/2197 since we are moving the overlays as well

d3zd3z reviewed (commented):
The change looks good. Once mcuboot has moved to a supporting version, I can give an approval.

danieldegrasse commented on west.yml:
```diff
@@ -23,6 +23,8 @@ manifest:
       url-base: https://github.com/zephyrproject-rtos
     - name: babblesim
       url-base: https://github.com/BabbleSim
+    - name: mcuboot
```
Great- I think we will need to be careful with synchronization here- the MCUBoot side change moves some overlay files into the MCUBoot repo that were previously defined here (used for the ramload sample). So I think that these changes (or a subset) will be needed with the PR that bumps the mcuboot manifest revision.

danieldegrasse commented on modules/Kconfig.mcuboot:
```diff
@@ -211,6 +211,26 @@ config MCUBOOT_BOOTLOADER_MODE_RAM_LOAD
+	  from there. The image must be linked to execute from RAM, the address that it is copied
+	  to is specified using the load-addr argument when running imgtool.
+	  This option automatically selectes MCUBOOT_BOOTLOADER_NO_DOWNGRADE as it is not possible
+	  to swap back to older version of the application.
```
Yeah, that is true. Perhaps

> This option automatically selectes MCUBOOT_BOOTLOADER_NO_DOWNGRADE as MCUBoot will automatically select the highest revision of the application to boot. Note however, MCUBoot will select an older revision of the application if the booted revision does not mark itself as confirmed.

danieldegrasse commented:
Updated now that Zephyr's copy of MCUBoot includes the relevant ramload support

danieldegrasse commented on share/sysbuild/image_configurations/MAIN_image_default.cmake:
```diff
@@ -39,6 +39,11 @@ if(SB_CONFIG_BOOTLOADER_MCUBOOT)
     set_config_bool(${ZCMAKE_APPLICATION} CONFIG_XIP n)
     set_config_int(${ZCMAKE_APPLICATION} CONFIG_FLASH_SIZE 0)
+  elseif(SB_CONFIG_MCUBOOT_MODE_RAM_LOAD_WITH_REVERT)
+    # RAM load mode requires XIP be disabled and flash size be set to 0
```
I don't, no- not familiar with `SINGLE_APP_RAM_LOAD`. @nordicjm do you know why this is the case?

nordicjm replied:
An oversight probably

nordicjm replied:
@edersondisouza

edersondisouza replied:
XIP is orthogonal to `SINGLE_APP_RAM_LOAD` - it simply loads the image to RAM. If the image was linked to run from the area in the RAM it is going to be loaded, it can use XIP.

danieldegrasse commented on samples/subsys/mgmt/mcumgr/smp_svr/sample.yaml:
```diff
@@ -183,8 +183,16 @@ tests:
+    platform_allow:
+      - mimxrt1050_evk/mimxrt1052/hyperflash
     integration_platforms:
-      - nrf52840dk/nrf52840
```
I'm not- the diff just looks weird. I'm adding a new testcase, `sample.mcumgr.smp_svr.ram_load_revert.serial`, and making the RT1050 the integration platform for that.

It is more obvious if you look at the file directly: https://github.com/danieldegrasse/zephyr/blob/feature/mcuboot-ramload-revert/samples/subsys/mgmt/mcumgr/smp_svr/sample.yaml#L179

maass-hamburg commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -96,15 +107,59 @@ uint8_t boot_fetch_active_slot(void)
 	}
 
 	LOG_DBG("Active slot: %d", slot);
+	/* Map slot number back to flash area ID */
+	switch (slot) {
+	case 0:
+		fa_id = FIXED_PARTITION_ID(SLOT0_PARTITION);
+		break;
+
+#if FIXED_PARTITION_EXISTS(SLOT1_PARTITION)
+	case 1:
+		fa_id = FIXED_PARTITION_ID(SLOT1_PARTITION);
```
this change is a breaking change, if people implemented the former behavior in their own compatible with mcuboot bootloader, should have gotten a option to preserve previous behavior, as changing this in their bootloader and getting that fixed bootloader on already shipped products, is often not possible, as bootloaders are not meant to be upgraded

danieldegrasse replied:
I guess you had some logic to map the flash area ID in your application? Could you use `BLINFO_BOOTLOADER_VERSION` to determine which version of MCUBoot is executing on your platform? Failing that, I have no issue with a Kconfig to use the previous functionality here

maass-hamburg commented on cmake/mcuboot.cmake:
```diff
@@ -130,7 +130,8 @@ function(zephyr_mcuboot_tasks)
     dt_reg_addr(slot1_partition_address PATH ${slot1_partition})
 
-    set(imgtool_args --align 1 --load-addr ${chosen_ram_address} ${imgtool_args})
+    dt_prop(write_block_size PATH "${flash_node}" PROPERTY "write-block-size")
+    set(imgtool_args --align ${write_block_size} --load-addr ${chosen_ram_address} ${imgtool_args})
```
fallback needed if write-block-size is not set. the generic spi-nor flash driver does not have this prop and also not allow this prop to be added. If it is not set imgtool will complain, that align is without argument

danieldegrasse replied:
I'd argue we need to allow the generic spi-nor-flash driver to add that property in this case- typically this is done by including the `soc-nv-flash` compatible, but you can also define the property directly

maass-hamburg replied:
added it here: https://github.com/zephyrproject-rtos/zephyr/pull/95152

erwango replied:
@danieldegrasse This change indeed broke the RAMLOAD build on our side.
I see #95152 is supposed to fix it, but it seems blocked for the time being. Would it be possible to revert this specific change waiting to have a consensus on a way to avoid breaking existing users ?
